jgrapht-core/src/main/java/org/jgrapht/alg/tour/DirectedHamiltonianCycle.java

```diff
@@ -0,0 +1,195 @@
+ * http://www.eclipse.org/legal/epl-2.0, or the
+ * GNU Lesser General Public License v2.1 or later
+ * which is available at
+ * http://www.gnu.org/licenses/old-licenses/lgpl-2.1-standalone.html.
+ *
+ * SPDX-License-Identifier: EPL-2.0 OR LGPL-2.1-or-later
+ */
+package org.jgrapht.alg.tour;
+
+import org.jgrapht.Graph;
+import org.jgrapht.GraphPath;
+import org.jgrapht.GraphTests;
+import org.jgrapht.alg.connectivity.KosarajuStrongConnectivityInspector;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+
+/**
+ * Simple algorithm for computing Hamiltonian cycles in directed graphs as described in
+ * <a href="https://youtu.be/dQr4wZCiJJ4">this YouTube video</a>. The algorithm's pseudo code can be seen starting from
+ * <a href="https://youtu.be/dQr4wZCiJJ4?t=348">time index 5:48</a>. Runtime grows exponentially(!) with the number of
+ * vertices in the graph.
+ * <p></p>
+ * A Hamiltonian cycle, also called a Hamiltonian circuit, Hamilton cycle, or Hamilton circuit, is a
+ * graph cycle (i.e., closed loop) through a graph that visits each node exactly once (Skiena 1990,
+ * p. 196).
+ *
+ * @param <V> the graph vertex type
+ * @param <E> the graph edge type
+ * @author Alexander Kriegisch
+ * @see <a href="https://en.wikipedia.org/wiki/Hamiltonian_path">Wikipedia</a> for a definition of Hamiltonian path and
+ * Hamiltonian cycle
+ */
+public class DirectedHamiltonianCycle<V, E> extends HamiltonianCycleAlgorithmBase<V, E> {
+  private Graph<V, E> graph;
+  private boolean cycleFound;
+  private List<V> vertices;
+  private int vertexCount;
+  private int maxIndex;
+  private int[] tourIndices;
+
+  // Implementation note: For whatever reason, int[] outperforms boolean[] for test graphs with several 100 vertices
+  // and several 1,000 edges. It outperforms BitSet even more dramatically. This might change for very large graphs.
+  // Background info about boolean[] vs. BitSet: https://www.baeldung.com/java-boolean-array-bitset-performance
+  private int[] adjacencyMatrix;
+
+  /**
+   * Computes a Hamiltonian tour
+   *
+   * @param graph input graph which must be directed, unweighted, have 3+ vertices, not allow self-loops and not allow
+   *              multiple edges
+   * @return Hamiltonian cycle (HC) if the graph is strongly connected, null otherwise. The HC is represented as a
+   * {@code GraphPath} in which the first and last vertices are identical, i.e. size is number of graph vertices + 1.
+   * @throws IllegalArgumentException if graph is undirected
+   * @throws IllegalArgumentException if graph contains &lt; 3 vertices (cycle impossible)
+   * @throws IllegalArgumentException if graph allows self-loops
+   * @throws IllegalArgumentException if graph is weighted
+   * @throws IllegalArgumentException if graph allows multiple edges between two vertices
+   */
+  @Override
+  public synchronized GraphPath<V, E> getTour(Graph<V, E> graph) {
```

Comment:
We don't put synchronized on algorithm methods; the current pattern throughout our library is that algorithm objects can only safely be accessed by one thread at a time.

Reply:
Oh, I see. But that requires sacred knowledge for first-time users. Feel free to remove the keyword after accepting the PR. Or is it a precondition for the code being accepted?

Reply:
We don't merge PR's until all review comments have been resolved.  There's nothing sacred here; one of the main purposes of code review is transmission of knowledge (sometimes it's tribal, sometimes it's already documented).

Reply:
I am not talking about transmission of knowledge between maintainers and contributors, rather about deliberately not making classes thread-safe, possibly causing avoidable problems for users knowing little about the product. Then when they have problems, of course you can tell them "it is documented", but I think avoiding or removing deficiencies of libraries is more prudent than documenting them. Anyway, I will remove `synchronized` both here and in class `CyclicTransitiveReduction`.